src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerListener.cs

```diff
@@ -62,17 +62,26 @@ public void Dispose()
 
         public async Task StartAsync(CancellationToken cancellationToken)
         {
-            if (this._listenerStarted)
+            if (this._listenerStatus == RegistrationStatus.Registered)
             {
                 throw new InvalidOperationException("The listener has already been started.");
             }
 
+            if (this._listenerStatus == RegistrationStatus.Registering)
+            {
+                throw new InvalidOperationException("The listener is starting.");
+            }
+
+            this._listenerStatus = RegistrationStatus.Registering;
             this.InitializeHost();
 
             try
             {
                 await this._host.RegisterObserverFactoryAsync(this);
-                this._listenerStarted = true;
+                if (this._listenerStatus == RegistrationStatus.Registering)
+                {
+                    this._listenerStatus = RegistrationStatus.Registered;
+                }
```

Comment:
There's a potential race condition here. Can you use `Interlocked.CompareExchange()` instead? Note that you'll have to get rid of the enum, I believe, and do something like we've done elsewhere with `int`s. Example here: https://github.com/Azure/azure-webjobs-sdk/blob/bb0b0e305ebbcb192cd57a3e6154c321a682e06a/src/Microsoft.Azure.WebJobs.Host/JobHost.cs#L94-L97.

Reply:
Note that you may want to do something similar above where you set the state to Registering (which would ensure that the state is `NotRegistered` before you begin).

Reply:
Done